File: 40-CombinationSumIi/40-CombinationSumIi.cpp
// Last updated: 04/04/2026, 23:38:32
class Solution {
public:
    void helper(vector<int>& arr,int idx,vector<int>& currList,int       target,vector<vector<int>>& ansList) {
        if (target == 0) {
            ansList.push_back(currList);
            return;
        }
        for (int i=idx; i<arr.size();i++) {
            if(i>idx && arr[i]==arr[i - 1]) continue;
            if(arr[i]>target) break;

            currList.push_back(arr[i]);
            helper(arr,i+1,currList,target-arr[i],ansList);
            currList.pop_back();
        }
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        vector<vector<int>> ans;
        vector<int> curr;
        helper(candidates,0,curr,target,ans);
        return ans;
    }
};

File: 40-CombinationSumIi/40-CombinationSumIi.java
// Last updated: 04/04/2026, 23:38:28
class Solution {
    private List<List<Integer>> ans = new ArrayList<>();
    private void backtrack(int idx, List<Integer> list, int target, int[] candidates) {
        if(target == 0) {
            ans.add(new ArrayList<>(list));
            return;
        }
        if(idx == candidates.length || target < 0) {
            return;
        }
        for(int i=idx; i<candidates.length;i++) {
            if(i>idx && candidates[i] == candidates[i-1]) continue;
            if(candidates[i]>target) break;
            list.add(candidates[i]);
            backtrack(i+1, list, target-candidates[i], candidates);
            list.remove(list.size()-1);
        }
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<Integer> list = new ArrayList<>();
        backtrack(0, list, target, candidates);
        return ans;
    }
}